### PR TITLE
feat(fastify.js): use caseSensitive option in default querystring-parser

### PR DESCRIPTION
**Pull request by Nikodermus**
Include a lowercase formatting when the case sensitive option is disabled

This should solve #2644 

hacktoberfest

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -88,8 +88,13 @@ function fastify (options) {
 
   validateBodyLimitOption(options.bodyLimit)
 
+  const caseSensitiveParse = (string) => {
+    const shouldLower = options.caseSensitive === false
+    return querystring.parse(shouldLower ? string.toLowerCase() : string)
+  }
```
Comment (mcollina) [inline fastify.js:94]: Can you please move this to a top level function?

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you please add a unit test?

File: fastify.js [hunk 2]
```diff
@@ -88,8 +88,13 @@ function fastify (options) {
 
   validateBodyLimitOption(options.bodyLimit)
 
+  const caseSensitiveParse = (string) => {
+    const shouldLower = options.caseSensitive === false
+    return querystring.parse(shouldLower ? string.toLowerCase() : string)
```
Comment (L2jLiga) [inline fastify.js:93]: I'm a bit worried, is it expected that value will be lowered as well as key?

```
> require('querystring').parse('name=Anr&nAmE=LL&namE=Noda'.toLowerCase())
[Object: null prototype] { name: [ 'anr', 'll', 'noda' ] }
```

Comment (zekth) [reply fastify.js:93]: Good catch, we don't want this

Comment (L2jLiga) [reply fastify.js:93]: So what is option here then?

Comment (zekth) [reply fastify.js:93]: solution would be something like: 
```js
const caseSensitiveParse = (string) => {
  const shouldLower = options.caseSensitive === false
  let out = querystring.parse(string)
  if(shouldLower) {
    out = Object.keys(out).reduce((acc, key) => { acc[key.toLowerCase()] = out[key]; return acc }, {})
  }
  return out
}
```

File: fastify.js [hunk 3]
```diff
@@ -88,8 +88,13 @@ function fastify (options) {
...
+  }
+
   const requestIdHeader = options.requestIdHeader || defaultInitOptions.requestIdHeader
-  const querystringParser = options.querystringParser || querystring.parse
+  const querystringParser = options.querystringParser || caseSensitiveParse
```
Comment (mcollina) [inline fastify.js:97]: Please move the from within the function to the initialization of fastify.

File: fastify.js (hunk 2 above)
Comment (L2jLiga) [reply fastify.js:93]: Yeah, but it does not take in account cases when query is `name=1&name=2` for example

Comment (L2jLiga) [reply fastify.js:93]: So seems like we have to make our own solution and cover all the places to add option for case-insensetive query parsing

File: fastify.js (hunk 1 above)
Comment (Nikodermus) [reply fastify.js:94]: @mcollina I'm trying to do a unit test this way, but I get a timeout error, could you guide me what's going here? I made if off the `custom-querystring-parser.test.js`

```
// default-querystring-parser.test.js
'use strict'

const t = require('tap')
const test = t.test
const sget = require('simple-get').concat
const Fastify = require('..')

test('Default querystring parser should keep case sensitive keys', t => {
  t.plan(9)

  const fastify = Fastify()

  fastify.get('/', (req, reply) => {
    t.deepEqual(req.query, {
      Foo: 'bar',
      Baz: 'faz'
    })
    reply.send({ hello: 'world' })
  })

  fastify.listen(0, (err, address) => {
    t.error(err)
    t.tearDown(() => fastify.close())

    sget({
      method: 'GET',
      url: `${address}?Foo=bar&Baz=faz`
    }, (err, response, body) => {
      t.error(err)
      t.strictEqual(response.statusCode, 200)
    })

    fastify.inject({
      method: 'GET',
      url: `${address}?Foo=bar&Baz=faz`
    }, (err, response, body) => {
      t.error(err)
      t.strictEqual(response.statusCode, 200)
    })
  })
})
```

File: fastify.js (hunk 3 above)
Comment (Nikodermus) [reply fastify.js:97]: Where is that? I tried to move it outside of the `fastify` function but I need access to `options` to know when to lowercase

File: fastify.js (hunk 2 above)
Comment (Nikodermus) [reply fastify.js:93]: What about this?

```
const caseSensitiveParse = (string) => {
    const shouldLower = options.caseSensitive === false
    let out = querystring.parse(string)

    if (shouldLower) {
      out = Object.keys(out).reduce((acc, key) => {
        const lowerKey = key.toLowerCase()
        const existingItem = acc[lowerKey]

        // If a variation of the key exists, merge both
        if (existingItem) {
          acc[lowerKey] = [existingItem, out[key]].flat()
        } else {
          acc[lowerKey] = out[key]
        }

        return acc
      },
      {})
    }
    return out
  }
```

Comment (zekth) [reply fastify.js:93]: > Yeah, but it does not take in account cases when query is `name=1&name=2` for example

so in the case of a querystring like `name=Anr&nAmE=LL&namE=Noda` the solution of @Nikodermus would work and output `name: ['Anr','LL','Noda']`

Comment (L2jLiga) [reply fastify.js:93]: I'm almost ok with that solution but we should take in account that fastify supports Node 10+ and Node 10 does not support [`Array.prototype.flat`](https://node.green/#ES2019-features-Array-prototype--flat--flatMap-)

File: fastify.js (hunk 1 above)
Comment (mcollina) [reply fastify.js:94]: I think the use of `address` is wrong with `sget`. That;s not a `url`.

**Comment by mcollina** [comment]
Any updates here?

File: fastify.js (hunk 2 above)
Comment (Nikodermus) [reply fastify.js]: Made an update, using reduce instead of .flat

File: fastify.js (hunk 1 above)
Comment (Nikodermus) [reply fastify.js]: Would you be able to put an example for me? I don't get how testing for apis work.

Here are some cases I have in mind:

- *`options.caseSensitive = true | undefined`:* `foo=bar&Foo=bar` => `{foo: 'bar', Foo: 'bar'}`
- *`options.caseSensitive = false`:* `foo=bar&Baz=bar` => `{foo: 'bar', baz: 'bar'}`
- *`options.caseSensitive = false`:* `foo=bar&Foo=Bar` => `{foo: ['bar','Bar']}`
- *`options.caseSensitive = true | undefined`:* `foo=bar&Foo=bar` => `{foo: 'bar', Foo: 'bar'}`

Comment (Nikodermus) [reply fastify.js]: nvm I found some examples, tho I'm having some issues with other tests, any chance you could look at @mcollina ?

File: docs/Server.md [hunk 4]
```diff
@@ -228,6 +228,8 @@ By default, value equal to `true`, routes are registered as case sensitive. That
 
 By setting `caseSensitive` to `false`, all paths will be matched as lowercase, but the route parameters or wildcards will maintain their original letter casing.
 
+This option also applies to query strings. When `caseSensitive` is set to `false`, `/foo?query=bar` is equivalent to `/foo?QUERY=bar` and the other variants (`qUerY`, `Query`, etc.). Properties with case variations are converted into a single argument. e.g. (`foo=bar1&Foo=bar2&foO=bar3` gets converted to `{foo: ['bar1', 'bar2', 'bar3']}`
```
Comment (L2jLiga) [inline docs/Server.md:231]: As @jsumners mentioned [here](https://github.com/fastify/fastify/pull/2649#discussion_r511913091) this is breaking change

**Comment by Nikodermus** [comment]
What's the next step then? Close this?

**Comment by L2jLiga** [comment]
I'm unsure, let's see what @mcollina and @jsumners say, btw I would like to make CI green by fixing `test/custom-querystring-parser.test.js`

File: fastify.js [hunk 5]
```diff
@@ -59,6 +59,28 @@ const onBadUrlContext = {
...
+  const original = querystring.parse(string)
+  const parsed = Object.keys(original).reduce((acc, key) => {
+    const lowerKey = key.toLowerCase()
+    const existingItem = acc[lowerKey]
+    const originalValue = original[key]
+
+    // If a variation of the key exists, merge both
+    if (existingItem) {
+      // Flat array
+      acc[lowerKey] = [existingItem, originalValue].reduce((acc, val) => acc.concat(val), [])
```
Comment (jsumners) [inline fastify.js:72]: This seems surprising and should be noted in the documentation. How is this different, if it is, from the way core `querystring.parse` handles same keys?

Comment (Nikodermus) [reply fastify.js:72]: Yes, it's working the same as the default core: https://www.npmjs.com/package/query-string
image

File: fastify.js [hunk 6]
```diff
@@ -59,6 +59,28 @@ const onBadUrlContext = {
   onError: []
 }
 
+const caseInsensitiveParse = (string) => {
```
Comment (salmanm) [inline fastify.js:62]: Do you guys think we can use `qs` when `caseSensitive` is `false`? Much simpler than the custom function.

```js
qs.parse(str, {
  decoder: (str, encoder, _, type) => type === 'key' ? encoder(str.toLowerCase()) : encoder(str)
})
```

_Although I agree with [this](https://github.com/fastify/fastify/issues/2644#issuecomment-715498064), simply using custom `querystringParser` is the recommended fix in this case._

Comment (L2jLiga) [reply fastify.js:62]: I wouldn't add dependencies but in general I'm ok

Comment (mcollina) [reply fastify.js:62]: I don't want to add qs to the deps. We might also just document if that somebody wants case insensitive query string keys.. then they can use qs. This is already possible, and maybe this PR is actually not needed at all.

Comment (Nikodermus) [reply fastify.js:62]: If that's the case, I can add that and close this one, just give me a confirmation

Comment (jsumners) [reply fastify.js:62]: Yeah, docs saying "if you need this, use `qs`" is the easier and better solution.

Comment (mcollina) [reply fastify.js:62]: Let's go for that.

**Pull request closed without merging**